Couchdb 2.3 unsupported due to clustered "purge_seq" field transition to string value

## Bug Description

Couchdb 2.3 changed the purge_seq field in the `DbInfo` response to be a string (this possibly applies to other sequences as well)

### 1. Steps to reproduce and the simplest code sample possible to demonstrate the issue

Try to configure a client against a CouchDB 2.3 server and issue and query that'll call the  `info()` method from a `Database` instance.

### 2. What you expected to happen

This should have worked

### 3. What actually happened

Gson deserialization failed with the following stacktrace:
```
Caused by: com.google.gson.JsonSyntaxException: java.lang.NumberFormatException: For input string: "0-g1AAAAFTeJzLYWBg4MhgTmEQTM4vTc5ISXIwNDLXMwBCwxygFFOSA5BMiv___39WIgMedXksQJLhAJACKr2PX22SAshMe4JmJjIkyRNpcQPE4vnEqH0AUUuUuQsgavcT8FACyEP1EDOzADZYXo4"
	at com.google.gson.internal.bind.TypeAdapters$11.read(TypeAdapters.java:306)
	at com.google.gson.internal.bind.TypeAdapters$11.read(TypeAdapters.java:296)
	at io.gsonfire.gson.ExcludeByValueTypeAdapterFactory$ExcludeByValueTypeAdapter.read(ExcludeByValueTypeAdapterFactory.java:99)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
	at io.gsonfire.gson.ExcludeByValueTypeAdapterFactory$ExcludeByValueTypeAdapter.read(ExcludeByValueTypeAdapterFactory.java:99)
	at io.gsonfire.gson.HooksTypeAdapter.deserialize(HooksTypeAdapter.java:86)
	at io.gsonfire.gson.HooksTypeAdapter.read(HooksTypeAdapter.java:54)
	at com.google.gson.Gson.fromJson(Gson.java:927)
	at com.google.gson.Gson.fromJson(Gson.java:865)
	at com.cloudant.client.org.lightcouch.internal.CouchDbUtil.getResponse(CouchDbUtil.java:229)
	at com.cloudant.client.org.lightcouch.CouchDbClient.get(CouchDbClient.java:384)
	at com.cloudant.client.api.Database.info(Database.java:1333)
```


## Environment details
- CouchDB versions
   > 2.3
- Version(s) that are affected by this issue.
   > 2.13.1
- Java version including vendor and platform
   > JDK 11 Linux (Azul) , but relevant to all platforms
- Using the optional okhttp dependency.
   > No

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,3 +1,7 @@
+# UNRELEASED
+- [NEW] Add `purge_seq` getter on `DbInfo` class that returns sequence value as `String`.
+- [DEPRECATED] `com.cloudant.client.api.model.DbInfo#getPurgeSeq()`.
+
 # 2.13.1 (2018-09-05)
 - [FIXED] Regression that prevented `keys` being included as part of a `ViewMultipleRequest`.
 

--- a/cloudant-client/src/main/java/com/cloudant/client/api/model/DbInfo.java
+++ b/cloudant-client/src/main/java/com/cloudant/client/api/model/DbInfo.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015 IBM Corp. All rights reserved.
+ * Copyright © 2015, 2018 IBM Corp. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
  * except in compliance with the License. You may obtain a copy of the License at
@@ -16,6 +16,7 @@ package com.cloudant.client.api.model;
 
 
 import com.google.gson.JsonElement;
+import com.google.gson.JsonPrimitive;
 import com.google.gson.annotations.SerializedName;
 
 /**
@@ -33,7 +34,7 @@ public class DbInfo {
     @SerializedName("update_seq")
     private JsonElement updateSeq;
     @SerializedName("purge_seq")
-    private long purgeSeq;
+    private JsonElement purgeSeq;
     @SerializedName("compact_running")
     private boolean compactRunning;
     @SerializedName("disk_size")
@@ -59,8 +60,37 @@ public class DbInfo {
         return updateSeq.toString();
     }
 
+    /**
+     * Use {@link #getStringPurgeSeq()} instead.
+     *
+     * The value 0 is returned if the {@code purged_seq} field cannot be cast as a primitive long.
+     * In later versions of CouchDB (&gt;=2.3.x) {@code purged_seq} is an opaque string.
+     *
+     * @return Number of purge operations on the database.
+     */
+    @Deprecated
     public long getPurgeSeq() {
-        return purgeSeq;
+        try {
+            JsonPrimitive purgeSeqPrim = purgeSeq.getAsJsonPrimitive();
+            return purgeSeqPrim.getAsLong();
+        } catch (IllegalStateException e) {
+            // Suppress exception if the element is of type JsonArray but contains more than a
+            // single element.
+        } catch (NumberFormatException e) {
+            // Suppress exception if the element is not a JsonPrimitive and is not a valid long
+            // value.
+        }
+        // Return 0 when value cannot be cast as a primitive long value.
+        return 0;
+    }
+
+    /**
+     * An opaque string that describes the state of purge operations across the database.
+     *
+     * @return Purge sequence.
+     */
+    public String getStringPurgeSeq() {
+        return purgeSeq.toString();
     }
 
     public boolean isCompactRunning() {

--- a/cloudant-client/src/test/java/com/cloudant/tests/DbInfoMockTests.java
+++ b/cloudant-client/src/test/java/com/cloudant/tests/DbInfoMockTests.java
@@ -1,0 +1,86 @@
+/*
+ * Copyright © 2018 IBM Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+package com.cloudant.tests;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import com.cloudant.client.api.CloudantClient;
+import com.cloudant.client.api.Database;
+import com.cloudant.tests.base.TestWithMockedServer;
+
+import org.junit.jupiter.api.Test;
+
+import okhttp3.mockwebserver.MockResponse;
+
+/**
+ * Created by samsmith on 13/12/2018.
+ */
+
+public class DbInfoMockTests extends TestWithMockedServer {
+
+    @Test
+    public void getDbInfoLongPurgeSeqAsLong() {
+        CloudantClient c = CloudantClientHelper.newMockWebServerClientBuilder(server).build();
+        Database db = c.database("animaldb", false);
+
+        MockResponse response = new MockResponse()
+                .setResponseCode(200)
+                .setBody("{\"purge_seq\":123}");
+        server.enqueue(response);
+
+        assertEquals(123, db.info().getPurgeSeq());
+    }
+
+    @Test
+    public void getDbInfoStringPurgeSeqAsLong() {
+        CloudantClient c = CloudantClientHelper.newMockWebServerClientBuilder(server).build();
+        Database db = c.database("animaldb", false);
+
+        String mockPurgeSeq = "\"3-g1AAAABXeJzLYWBgYMpgTmEQTM4vTc5ISXLIyU9OzMnILy7JAUklMiTV____PyuRAY-iPBYgydAApP6D1TJnAQCZtRxv\"";
+        MockResponse response = new MockResponse()
+                .setResponseCode(200)
+                .setBody("{\"purge_seq\":" + mockPurgeSeq + "}");
+        server.enqueue(response);
+
+        assertEquals(0, db.info().getPurgeSeq());
+    }
+
+    @Test
+    public void getDbInfoLongPurgeSeqAsString() {
+        CloudantClient c = CloudantClientHelper.newMockWebServerClientBuilder(server).build();
+        Database db = c.database("animaldb", false);
+
+        MockResponse response = new MockResponse()
+                .setResponseCode(200)
+                .setBody("{\"purge_seq\":123}");
+        server.enqueue(response);
+
+        assertEquals("123", db.info().getStringPurgeSeq());
+    }
+
+    @Test
+    public void getDbInfoStringPurgeSeqAsString() {
+        CloudantClient c = CloudantClientHelper.newMockWebServerClientBuilder(server).build();
+        Database db = c.database("animaldb", false);
+
+        String mockPurgeSeq = "\"3-g1AAAABXeJzLYWBgYMpgTmEQTM4vTc5ISXLIyU9OzMnILy7JAUklMiTV____PyuRAY-iPBYgydAApP6D1TJnAQCZtRxv\"";
+        MockResponse response = new MockResponse()
+                .setResponseCode(200)
+                .setBody("{\"purge_seq\":" + mockPurgeSeq + "}");
+        server.enqueue(response);
+
+        assertEquals(mockPurgeSeq, db.info().getStringPurgeSeq());
+    }
+}

